src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ISMTemplateService.kt

```diff
@@ -98,30 +139,368 @@ fun validateFormat(indexPatterns: List<String>): ElasticsearchException? {
+                            xContentRegistry,
+                            LoggingDeprecationHandler.INSTANCE,
+                            response.sourceAsBytesRef,
+                            XContentType.JSON
+                        ).use { xcp ->
+                            xcp.parseWithType(response.id, response.seqNo, response.primaryTerm, Policy.Companion::parse)
+                        }
+                    } catch (e: Exception) {
+                        logger.error("Failed to parse policy [${response.id}] when migrating templates", e)
+                    }
+
+                    if (policy?.ismTemplate == null) {
+                        policiesToUpdate[it.id] = Pair(response.seqNo, response.primaryTerm)
+                    }
+                }
+            }
+        } catch (e: ActionRequestValidationException) {
+            logger.warn("ISM config index not exists when migrating templates.")
+        }
+    }
+
+    private suspend fun updateISMPolicies() {
+        if (policiesToUpdate.isEmpty()) return
+
+        if (!imIndices.attemptUpdateConfigIndexMapping()) {
+            logger.error("Failed to update config index mapping.")
+            return
+        }
+
+        val requests = mutableListOf<UpdateRequest>()
+        policiesToUpdate.forEach { policyID, (seqNo, priTerm) ->
+            val ismTemplates = ismTemplateMap[policyID]
+            if (ismTemplates != null)
+                requests.add(updateISMTemplateRequest(policyID, ismTemplates, seqNo, priTerm))
+        }
+        var requestsToRetry: List<DocWriteRequest<*>> = requests
+
+        retryPolicy.retry(logger, listOf(RestStatus.TOO_MANY_REQUESTS)) {
+            val bulkReq = BulkRequest().add(requestsToRetry)
+            val bulkRes: BulkResponse = client.suspendUntil { bulk(bulkReq, it) }
+            val failedResponses = (bulkRes.items ?: arrayOf()).filter { it.isFailed }
+
+            val retryItems = mutableListOf<Int>()
+            bulkRes.items.forEach {
+                if (it.isFailed) {
+                    if (it.status() == RestStatus.TOO_MANY_REQUESTS) {
+                        retryItems.add(it.itemId)
+                    } else {
+                        logger.error("Failed to update template for policy [${it.id}], ${it.failureMessage}")
+                        policiesFailedToUpdate[it.id] = it.failure
+                    }
+                } else {
+                    policiesToUpdate.remove(it.id)
+                    policiesFailedToUpdate.remove(it.id)
+                }
+            }
+            requestsToRetry = retryItems.map { bulkReq.requests()[it] as UpdateRequest }
+
+            if (requestsToRetry.isNotEmpty()) {
+                val retryCause = failedResponses.first { it.status() == RestStatus.TOO_MANY_REQUESTS }.failure.cause
+                throw ExceptionsHelper.convertToElastic(retryCause)
+            }
+        }
+    }
+
+    private fun cleanCache() {
```

Comment:
I think in a few places we don't catch all exceptions or we throw exceptions during the migration process. This would mean the cache was not cleared and these in memory structures are filled with data. If that happens and then the migration runs again, will we get incorrect results? Should we clear the cache at the start of every migration attempt?

Reply:
Good point, we should clear cache at the beginning of every run.